File: 2016/day5.js
let md5 = require('./md5')

let input = 'ugkcyxxg'

let pwd = ''
let arr = new Array(8)
let i = 0
let cinema = '________'.split('')

function randomString () {
  return String.fromCharCode(33 + Math.floor(Math.random() * 100))
}
while (true) {
  let hash = md5(input + i)
  if (hash.match(/^0{5}/)) {
    let char = hash.charAt(5)
    //pwd += char
    if (char.match(/[0-7]/)) {
      let index = parseInt(char)
      arr[index] = arr[index] || hash.charAt(6)
      for (let j = 0; j < arr.length; j++) {
        cinema[j] = arr[j] || '_'
      }
      if (!cinema.includes('_')) {
        break
      }
    }
  }
  if (i % 100 === 0)
    console.log(cinema.join('').replace(/_/g, randomString))
  /*if (pwd.length === 8) {
    break
  }*/
  i++
}
console.log(cinema.join(''))

File: 2016/day5.rb
require 'digest'
input = "ugkcyxxp"

password = ''
i = 0
while password.length < 8 
  hash = Digest::MD5.hexdigest "#{input}#{i}"
  if hash.start_with? "00000"
    password += hash[5]
  end
  i += 1
end
puts password

password = []
i = 0
while password.find_all{|c| c}.length < 8
  hash = Digest::MD5.hexdigest "#{input}#{i}"
  if hash.start_with?("00000") && (0..7).map(&:to_s).include?(hash[5])
    password[hash[5].to_i] = hash[6]
  end
  puts password.map{|c| c || '_'}.join
  i += 1
end
puts password

